Introduce SecureMiningReward action

longfin opened the pull request:
This PR introduces `SecureMiningReward` to retrieve NCG from authorized miners to arbitrary addresses, for [NCIP-10](https://docs.google.com/document/d/1ErZ5JQia03KqXRG6IRZ7SORfnxMLZfJg4patVKFGX5Y/edit).

longfin commented on Lib9c/Action/SecureMiningReward.cs:
```diff
@@ -0,0 +1,65 @@
+
+            CheckPermission(context);
+
+            // TODO consider a better solution to determine.
```
```suggestion
            // FIXME considers better solutions to check.
```

longfin commented on Lib9c/Action/SecureMiningReward.cs:
```diff
@@ -0,0 +1,65 @@
+                throw new InvalidAddressException($"Given address[{Miner}] isn't an authorized miner");
+            }
+
+            // TODO Do we need the maximum cap per miner? (but... if then, how do we maintain?)
```
```suggestion
```

moreal commented on Lib9c/Action/SecureMiningReward.cs:
```diff
@@ -0,0 +1,73 @@
+            {
+                return state.MarkBalanceChanged(
+                    Amount.Currency,
+                    new [] { Miner, Recipient}
```
Just style comment, you can ignore.

```suggestion
                    new [] { Miner, Recipient }
```

moreal commented:
Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that *Planetarium* can do it well? Or just not considered?

moreal reviewed (commented):
Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that *Planetarium* can do it well? Or just not considered?

longfin commented:
> Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that _Planetarium_ can do it well? Or just not considered?

It is assumed that the user who is authorized (i.e., passes `CheckPermission()`) performs "well". 😅

longfin commented:
> > It is assumed that the user who is authorized (i.e., passes `CheckPermission()`) performs "well". 😅
> 
> I'm very careful to reply. But, in my honest opinion, I think it is better to automate by implementing the NCIP-10 as the action code, to block the possibilities of mistakes. It seems able to be implemented.
> 
> Of course, it is just my preferred way so if they prefer this way, I will not object.

@moreal First, thanks for your opinion and I also understand that this action seems quite strange.

- Despiting I've mentioned NCIP-10, it lacks some information to apply directly (e.g., it doesn't specify the actual starting block or miner address).
   - but on second thought, it had an address to withdraw (i.e., `0xB3bCa3b3c6069EF5Bdd6384bAD98F11378Dc360E`). so it might be better to force that address.
- Even if we fixed block indices arbitrarily, it isn't matched the actual amount for a specific miner because we've changed the addresses of the miner for reasons such as outages.

To sum up, if we secure all miner's rewards for all periods instead of NCIP-10's, it can be executed automatically and easily.

@jckdotim do you have any idea?

longfin commented:
Amended https://github.com/planetarium/lib9c/pull/1720/commits/2d2d018a7e0d057168b638c77f0ba3013038e1b3  to address https://github.com/planetarium/lib9c/pull/1720#issuecomment-1439915335. 

PTAL @moreal @jckdotim

jckdotim commented:
@longfin I agree with your suggestion to automate the process as the action code to prevent mistakes.
